parser.go

```diff
@@ -0,0 +1,345 @@
+				Message: err.Error(),
+			})
+			continue
+		}
+
+		if paramName == "-" || tag.paramSet {
+			continue
+		}
+
+		if !consts.ParamNameRE.MatchString(paramName) {
+			violations = append(violations, types.Violation{
+				Path:    member.Path,
+				SetName: setName,
+				Message: fmt.Sprintf("Name %q is invalid for parameter or set (valid: %s)",
+					paramName, consts.ParamNameRE)})
+		}
+
+		tag.path = member.Path
+		ret.fields[paramName] = tag
+	}
+
+	if len(violations) > 0 {
+		return ret, violations
+	}
+
+	return ret, nil
+}
+
+func parseParam(structField reflect.StructField, fieldVal reflect.Value) (
+	paramName string,
+	_ paramSetField,
+	_ error,
+) {
+	tagParam := structField.Tag.Get("param")
+	tagParamParts := strings.Split(tagParam, ",")
+	paramName = tagParamParts[0]
+
+	ret := paramSetField{
+		typ:  describeType(fieldVal),
+		desc: structField.Tag.Get("param_desc"),
+	}
+
+	for _, tagOption := range tagParamParts[1:] {
+		switch tagOption {
+		case "optional":
+			ret.optional = true
+		case "secret":
+			ret.secret = true
+		default:
+			return paramName, ret, fmt.Errorf(
+				"option '%s' is invalid for tag 'param' in '%s'",
+				tagOption,
+				tagParam)
+		}
+	}
+
+	// if the parameter name is not provided in the "param" tag field then
+	// the name of the struct member in lowercase is used as parameter
+	// name.
+	if paramName == "" {
+		paramName = strings.ToLower(structField.Name)
+	}
+
+	// try to configure it as a "basic type"
+	err := configStandardCallbacks(&ret, fieldVal)
+	if err == nil {
+		return paramName, ret, nil
+	}
+
+	// try to configure it as an "xtype"
+	if isXType(structField.Type) {
+		if fieldVal.IsNil() {
+			fieldVal.Set(reflect.New(fieldVal.Type().Elem()))
+		}
+
+		ret.boolean = describeType(fieldVal) == "bool"
+
+		ret.validFn = toXType(fieldVal).ValueValid
+		ret.setValueFn = toXType(fieldVal).UnmarshalParam
+		ret.getDefaultFn = toXType(fieldVal).GetDefaultValue
+
+		// some types know how to redact themselves (for example,
+		// xtype.URL know how to redact the password)
+		if redactor := toRedactor(fieldVal); redactor != nil {
+			ret.redactFn = redactor.RedactValue
+		} else {
+			ret.redactFn = func(s string) string { return s }
+		}
+
+		return paramName, ret, nil
+	}
+
+	// if is a struct, assume it to be a parameter set
+	if fieldVal.Kind() == reflect.Struct {
+		ret.paramSet = true
+
+		// parameter sets have no value, and the callback functions should
+		// not be called; install handlers to help debug in case of a mistake.
+		panicMessage := fmt.Sprintf("%q is a paramset, it have no value", paramName)
+		ret.validFn = func(v string) error { panic(panicMessage) }
+		ret.setValueFn = func(v *string) error { panic(panicMessage) }
+		ret.getDefaultFn = func() (string, error) { panic(panicMessage) }
+		ret.redactFn = func(s string) string { panic(panicMessage) }
+
+		return paramName, ret, nil
+	}
+
+	return paramName, ret, fmt.Errorf("struct member %q is unsupported", paramName)
```

Comment:
When I don't set the UpdateFN func I get this error messages.
The message does not make it clear in this case that UpdateFN was not set.

It would be better to have an error message that says something like "UpdateFn" function was not set`